subsys/dap/cmsis_dap.c

```diff
@@ -0,0 +1,1062 @@
+		if (request[1]) {
+			LOG_INF("Debugger connected");
+		} else {
+			LOG_INF("Debugger disconnected");
+		}
+		break;
+	case DAP_TARGET_RUNNING:
+		LOG_DBG("unsupported");
+		break;
+	default:
+		*response = DAP_ERROR;
+		return 1U;
+	}
+
+	response[0] = DAP_OK;
+	return 1U;
+}
+
+/* Process Connect command and prepare response */
+static uint16_t dap_connect(struct dap_context *const ctx,
+			    const uint8_t *const request,
+			    uint8_t *const response)
+{
+	const struct swdp_api *api = ctx->swdp_dev->api;
+	uint8_t port;
+
+	if (request[0] == DAP_PORT_AUTODETECT) {
+		port = DAP_PORT_SWD;
+	} else {
+		port = request[0];
+	}
+
+	switch (port) {
+	case DAP_PORT_SWD:
+		LOG_INF("port swd");
+		ctx->debug_port = DAP_PORT_SWD;
+
+		if (atomic_test_and_set_bit(&ctx->state,
+					    DAP_STATE_CONNECTED)) {
+			LOG_ERR("DAP device is already connected");
+			break;
+		}
+
+		api->swdp_port_on(ctx->swdp_dev);
+		break;
+	case DAP_PORT_JTAG:
+		LOG_ERR("port unsupported");
+		port = DAP_ERROR;
+		break;
+	default:
+		LOG_DBG("port disabled");
+		port = DAP_PORT_DISABLED;
+		break;
+	}
+
+	response[0] = port;
+	return 1U;
+}
+
+/* Process Disconnect command and prepare response */
+static uint16_t dap_disconnect(struct dap_context *const ctx,
+			       uint8_t *const response)
+{
+	const struct swdp_api *api = ctx->swdp_dev->api;
+
+	LOG_DBG("");
+
+	ctx->debug_port = DAP_PORT_DISABLED;
+
+	if (atomic_test_bit(&ctx->state, DAP_STATE_CONNECTED)) {
+		api->swdp_port_off(ctx->swdp_dev);
+	} else {
+		LOG_WRN("DAP device is not connected");
+	}
+
+	response[0] = DAP_OK;
+	atomic_clear_bit(&ctx->state, DAP_STATE_CONNECTED);
+
+	return 1;
+}
+
+/* Process Delay command and prepare response */
+static uint16_t dap_delay(struct dap_context *const ctx,
+			  const uint8_t *const request,
+			  uint8_t *const response)
+{
+	uint16_t delay = sys_get_le16(&request[0]);
+
+	LOG_DBG("dap delay %u ms", delay);
+
+	k_busy_wait(delay * USEC_PER_MSEC);
+	response[0] = DAP_OK;
+
+	return 1U;
+}
+
+/* Process Reset Target command and prepare response */
+static uint16_t dap_reset_target(struct dap_context *const ctx,
+				 uint8_t *const response)
+{
+	response[0] = DAP_OK;
+	response[1] = 0U;
+	LOG_WRN("unsupported");
+
+	return 2;
+}
+
+/* Process SWJ Pins command and prepare response */
+static uint16_t dap_swj_pins(struct dap_context *const ctx,
+			     const uint8_t *const request,
+			     uint8_t *const response)
+{
+	const struct swdp_api *api = ctx->swdp_dev->api;
+	uint8_t value = request[0];
+	uint8_t select = request[1];
+	uint32_t wait = sys_get_le32(&request[2]);
+	int64_t timeout_ticks;
+	uint8_t state;
+
+	if (!atomic_test_bit(&ctx->state, DAP_STATE_CONNECTED)) {
+		LOG_ERR("DAP device is not connected");
+		response[0] = DAP_ERROR;
+		return 1U;
+	}
+
+	/* Skip if nothing selected. */
+	if (select) {
+		api->swdp_set_pins(ctx->swdp_dev, select, value);
+	}
+
+	timeout_ticks = k_uptime_ticks() + (CONFIG_SYS_CLOCK_TICKS_PER_SEC / 1000000 * wait);
```

Comment:
```suggestion
	timeout_ticks = k_uptime_ticks() + (CONFIG_SYS_CLOCK_TICKS_PER_SEC * wait / 1000000);
```
When `CONFIG_SYS_CLOCK_TICKS_PER_SEC=32768` dividing it by `1000000` will always result in 0. We should first multiply, then divide.

Reply:
Maybe this can be calculated using the `k_timepoint_t` instead?

Reply:
Great idea! I was not aware of this API :)